protocol/chainlib/provider_node_subscription_manager.go

```diff
@@ -0,0 +1,520 @@
+	return nil
+}
+
+func (pnsm *ProviderNodeSubscriptionManager) handleNewNodeMessage(ctx context.Context, hashedParams string, nodeMsg interface{}) {
+	pnsm.lock.RLock()
+	defer pnsm.lock.RUnlock()
+
+	if _, ok := pnsm.activeSubscriptions[hashedParams]; !ok {
+		utils.LavaFormatDebug("No hashed params in handleNewNodeMessage, connection might have been closed", utils.LogAttr("hash", hashedParams))
+		return
+	}
+	// Sending message to all connected consumers
+	for consumerAddrString, connectedConsumerContainer := range pnsm.activeSubscriptions[hashedParams].connectedConsumers {
+		utils.LavaFormatTrace("ProviderNodeSubscriptionManager:startListeningForSubscription() sending to consumer", utils.LogAttr("consumerAddr", consumerAddrString), utils.LogAttr("hashedParams", utils.ToHexString(hashedParams)))
+
+		copiedRequest := &pairingtypes.RelayRequest{}
+		copyRequestErr := protocopy.DeepCopyProtoObject(connectedConsumerContainer.firstSetupRequest, copiedRequest)
+		if copyRequestErr != nil {
+			utils.LavaFormatError("failed to copy subscription request", copyRequestErr)
+			return
+		}
+
+		extensionInfo := extensionslib.ExtensionInfo{LatestBlock: 0, ExtensionOverride: copiedRequest.RelayData.Extensions}
+		if extensionInfo.ExtensionOverride == nil { // in case consumer did not set an extension, we skip the extension parsing and we are sending it to the regular url
+			extensionInfo.ExtensionOverride = []string{}
+		}
+
+		chainMessage, err := pnsm.chainParser.ParseMsg(copiedRequest.RelayData.ApiUrl, copiedRequest.RelayData.Data, copiedRequest.RelayData.ConnectionType, copiedRequest.RelayData.GetMetadata(), extensionInfo)
+		if err != nil {
+			utils.LavaFormatError("failed to parse message", err)
+			return
+		}
+
+		apiCollection := pnsm.activeSubscriptions[hashedParams].apiCollection
+
+		marshalledNodeMsg, err := pnsm.convertNodeMsgToMarshalledJsonRpcResponse(nodeMsg, apiCollection)
+		if err != nil {
+			utils.LavaFormatError("error converting node message", err)
+			return
+		}
+
+		relayMessageFromNode := &pairingtypes.RelayReply{
+			Data:     marshalledNodeMsg,
+			Metadata: []pairingtypes.Metadata{},
+		}
+
+		err = pnsm.signReply(ctx, relayMessageFromNode, connectedConsumerContainer.consumerSDKAddress, chainMessage, copiedRequest)
+		if err != nil {
+			utils.LavaFormatError("error signing reply", err)
+			return
+		}
+
+		utils.LavaFormatDebug("Sending relay to consumer",
+			utils.LogAttr("requestRelayData", copiedRequest.RelayData),
+			utils.LogAttr("reply", marshalledNodeMsg),
+			utils.LogAttr("replyLatestBlock", relayMessageFromNode.LatestBlock),
+			utils.LogAttr("consumerAddr", connectedConsumerContainer.consumerSDKAddress),
+		)
+
+		connectedConsumerContainer.consumerChannel.Send(relayMessageFromNode)
+	}
+}
+
+func (pnsm *ProviderNodeSubscriptionManager) RemoveConsumer(ctx context.Context, chainMessage ChainMessageForSend, consumerAddr sdk.AccAddress, closeConsumerChannel bool) error {
```

Comment:
we need a consumer identifier, we can;t disconnect all consumer services talking to us that share the same address but on different servers, they might still have customers connected